lib/units/validateBody.js

```diff
@@ -148,6 +148,27 @@ function getBodyValidator(realType, expectedType) {
   return [null, validator[0]];
 }
 
+/**
+ * Parses a given JSON Schema to ensure it's a valid JSON.
+ * @param {string} schema
+ */
+const validateJsonSchema = (schema) => {
+  try {
+    JSON.parse(schema);
```

Comment:
I think at other places in Gavel we use `json-parse-helpfulerror` to provide even more helpful error messages than what `JSON.parse()` provides

Reply:
I was looking for how the message body was done, didn't find it with `JSON.parse`.

```
src/mixins/validatable-http-message.coffee:jph = require 'json-parse-helpfulerror'
src/mixins/validatable-http-message.coffee:        jph.parse @body
src/mixins/validatable-http-message.coffee:        jph.parse @body
src/mixins/validatable-http-message.coffee:          parsed = jph.parse @expected.bodySchema
src/mixins/validatable-http-message.coffee:          jph.parse @expected.body
src/mixins/validatable-http-message.coffee:          jph.parse @expected.body
```

Reply:
While looking over the changes yesterday I've spotted that we _already have_ a part of the body validation that attempts to parse a JSON Schema:

https://github.com/apiaryio/gavel.js/blob/bba82063cd9a7c1df1421a0764ddcd1b5ffd1cfd/lib/units/validateBody.js#L102-L109

Which means that mentioned issue scenario somehow doesn't trigger this logic. I would need to investigate. Hopefully, we don't need any new functions/checks, just need to fix the old ones.

Reply:
So, it looks like with the current founding we must treat malformed JSON Schemas like an exception. Shall we change this to an exception then?

```js
const error = `Can't validate: expected body JSON Schema is not a parseable JSON:\n${exception.message}`; 
  
   return [error, null]; 
```

> At the moment it's being tolerated as a validation error, not stopping the process, and ending up in the public results.

Reply:
I confirm that changing that soft handling to throwing an exception produces the same behavior:

* Stops Gavel validation to prevent operating on malformed data
* Propagates user-friendly error + JSON parsing error from `jhp`

Reply:
I think exception is all right. If Gavel is given schema, the user's intention is to validate according to it. If Gavel cannot parse it, then it shouldn't be silent or just warning. There's no way for Gavel to continue in that case, otherwise it could lead to false positives in validation.